clarify JSON API connection (#137)
Materials-Consortia/OPTiMaDe: optimade.md

> # OPTiMaDe API specification v0.9.8-develop

Reviewer: Shouldn't there be a new line after a header in markdown?

Author: This change was not intentional. I've put it back.

* **Database-provider-specific prefix**: This specification defines a set of
  database-provider-specific prefixes in [Appendix 1](#h.app1).
* **Implementation**: An instance serving the OPTiMaDe API.

> * **JSON API response schema**: Schema for the default response format, which MUST be *compatible* with the [JSON API v1.0](http://jsonapi.org/format/1.0) (see also [3.3 Responses](#h.3.3)).

Reviewer 2: Is the definition of the term JSON API response schema the right place to insert a side remark that the default response format MUST be compatible with jsonapi? Because at a first quick read I was confused: it sounds as if the sentence is trying to say that a JSON API response schema itself must be compatible with the jsonapi specification somehow. I'd like to see this side remark moved (integrated with what is already there) to 3.3.

Author: This is not a definition of the default response format. 
It is an explanation of what is meant by "JSON API response schema", which is used as a subtitle in many sections of the document (e.g. 4.1.2, 4.2.2 etc.)
This title can be confusing, since it can seem to imply that what follows below the title is something exclusively related to JSON API (while actually what follows will be a description of the default response format).

However, I agree with @rartino that this confusion would be better resolved by replacing all of these headlines with "JSON Response Schema" (or even "Default Response Schema"), in which case this term definition is no longer needed.
@rartino Should I go ahead and make the change?

Reviewer 2: I think I'm still confused about what these edits try to do, sorry.

A "JSON API response" is one out of several possible response formats. The "JSON API response schema" is a 'schema'-type description of the expectations of a response delivered in that format. Response formats are discussed in Sec 3, and there is some general general text about them in Section 3.3.1. As it seems you are concluding at the end in your reply, that means "JSON API response" / "JSON API Response Schema"  isn't something one would normally put as a **term definition**, just like we don't have a **term definition** of `chemical_formula_Hill`, but rather place all property definitions in a section of its own.

Now, with that background I don't quite understand the proposed rename: "JSON API response schema" -> "JSON Response Schema". If we in section 3.3.1 say that the default format is based on the "JSON API specification" and then refer back to that in headings and elsewhere as a "JSON API Response". Doesn't the headings "JSON API Resposne schema" make sense as written?

Author: @rartino It reads like you didn't notice that my changes removed the term definition?
Did you refresh the "Files" tab?

> Doesn't the headings "JSON API Resposne schema" make sense as written?

I don't think so because it is confusing as mentioned in my comment above.
The schema we provide there is a JSON schema for the response that we expect when specifying `response_format='json'`.

Reviewer 2: @ltalirz I see them; but given your suggested rename, I was still not sure if we agree on the reason for why to do that.

> The schema we provide there is a JSON schema for the response that we expect when specifying response_format='json'.

Right now, Section 4.1.1. specifies that the string is `response_format=jsonapi`...

"JSON" is just a serialized data format for general data. In the specification, as it is written right now, we thus talk about the "JSON API response format" to be more specific than "JSON response format". In that case, I'd also expect section titles to be, e.g., "JSON API Response schema" rather than "JSON Response schema", etc.

I suppose it would be OK to consistently drop the "API" everywhere. But then all mentions of "JSON API Response format", "JSON API format" etc. throughout the text needs to be edited to drop the "API". And the selection string should be ` response_format=json.`

(But I also don't think what we have right now is in error, or even confusing. It is just using a slightly more specific name for the default output format.)

Author: > Right now, Section 4.1.1. specifies that the string is response_format=jsonapi...

Ah - now I start to understand your comments better... I saw your previous comment on 
`jsonapi` vs `json` but did not realize that the **only** definition of the format string in the text was indeed `jsonapi` - sorry.
Do you know whether there is agreement on whether `json` and `jsonapi` should be used?

I would personally vote for `json` since it is something you can type in the URL and `json` might be more familiar to users of optimade (most of whom will have never heard of JSON API).

Reviewer 2: I wouldn't go as far as to say there is an agreement. But I think at some point, some of us, during the switch-over from the older json BUT NOT json api format into the json api format agreed to name the new format "jsonapi" to avoid confusion in the transition period, "json" would remain an obsolete reference to the old format.

BUT, with the latest major overhaul of the spec, I think that point is now moot. I do not feel strongly against overhauling the spec to consistently call the format "JSON resonse format".

Author: Alright, will do!

### <a name="h.3.3.1">3.3.1. Response Format</a>

By default, all API responses MUST comply with the [JSON API v1.0](http://jsonapi.org/format/1.0) specification.
In particular, all endpoints MUST be able to respond in the JSON API format.

> Each endpoint MAY support additional formats, and declare these formats in its own `info` endpoint

Reviewer 2: ```suggestion
Each endpoint MAY support additional formats, but these formats SHOULD be declared in the corresponding entry `info` endpoint
```
- Add a conformity keyword to the requirement (I'm going with SHOULD here as I don't want database-specific-prefixed formats temporarily being added for testing to formally break the spec.)
- Slight rephrase to clarify which info endpoint is meant.

(see section [4.3.2. Entry Listing Info Endpoints](#h.4.3.2)).
Clients can request these formats using the `response_format` URL query parameter (see below).

> Specifying a `response_format` URL query parameter different from `json`

Reviewer 2: Oh, wow, I see now that we are inconsistent throughout the spec whether the magical string is 'json' or 'jsonapi'. (Text says 'jsonapi', examples of info output uses 'json'). This needs to be fixed (not necessarily in this PR.)

Author: Agreed, see my comment above.

allows the API to break conformance with the JSON API specification not only
in response format but also, e.g., in terms of how content negotiation is implemented.

> API implementations MAY support additional formats beyond the ones specified in this document.

Reviewer 2: I'm confused; it seems the changes in this PR make this piece of text say the same thing twice, i.e. "Each endpoint MAY support additional formats" and "API implementations MAY support additional formats"? Can you try to integrate everything into one sentence?

Author: Well... the distinction here is between endpoint-specific additional formats (that are nevertheless specified in this document) and provider-specific additional formats (which are not specified in this document).

But I agree that this can be shortened. Will do.

In this case, the corresponding `response_format` identifiers MUST be include a

> database-provider-specific prefix as defined in [Appendix 1](#h.app1).

Reviewer 2: I don't like that this PR drops this text:
*"Specifying a `response_format` URL query parameter different from JSON API, allows the implementation to break conformance with the JSON API specification. Not only in response format, but also in, e.g., how content negotiation is implemented."*
This clarification seems quite crucial to me, in relation to the requirements of the jsonapi spec.

### <a name="h.3.3.1">3.3.1. Response Format</a>

This document defines a `json` response format that complies with the [JSON
API v1.0](http://jsonapi.org/format/1.0) specification.
All endpoints of an API implementation MUST be able to provide responses in the `json` format specified below
and MUST respond in this format by default.

> Each endpoint MAY support additional formats, and SHOULD declare these formats under `endpoint/info` 

Reviewer: To keep with convention, please encapsulate "endpoint" in your URI example with <> to show it is a placeholder.

(see section [4.3.2. Entry Listing Info Endpoints](#h.4.3.2)).
Clients can request these formats using the `response_format` URL query parameter.
Specifying a `response_format` different from `json` (e.g.
`response_format=xml`) allows the API to break conformance not only with the

> `json` response format specification, but also, e.g., in terms of how content

Reviewer: Maybe it should here be JSON instead of \`json\`, or?

negotiation is implemented.

> Database-provider-specific `response_format` identifiers MUST be include a

Reviewer: remove "be"

The API implementation MAY describe many-to-many relationships between entries along with OPTIONAL human-readable descriptions that describe each relationship. These relationships can be to the same, or to different, entry types.

> In responses that use the `json` response format, such relationships MUST be communicated using [JSON API Relationships](https://jsonapi.org/format/1.0/#document-resource-object-relationships) encoded in the `"relationships"` field of the response. The OPTIONAL human-readable description is provided in the `"description"` field inside the `"meta"` dictionary of a relationships object.

Reviewer: So, you are making it convention to write \`json\` instead of JSON, even though this is "only" the value to be queried for, when using `response_format`, while the "name" of the format is still JSON?

> URL tokens, and the above example is not to be taken as a recommendation for
> such a mechanism.

### <a name="h.4.1.2">4.1.2. JSON Response Schema</a>

"Entry listing" endpoint response dictionaries MUST have a `data`
key. The value of this key MUST be a list containing dictionaries that

> represent individual entries. In the default `json` response format every dictionary

Reviewer: Here, I would say JSON again, but this should be discussed in the first comment. I will, however, keep stating throughout the review, where I think it makes more sense to replace \`json\` with JSON, if you agree with me in the first place :)

A client can request a specific entry by appending an ID component to the URL of an entry listing
endpoint. This will return properties for the entry with that ID.

> In the default `json` response format, the ID component MUST be the content of the `id` field.

Reviewer: JSON

are OPTIONAL for clients, API implementations MUST accept and handle them:
**response\_format**, **email\_address**, **response\_fields**. The meaning of these URL query parameters are as defined above in section [4.1.1. URL Query Parameters](#h.4.1.1).

### <a name="h.4.2.2">4.2.2. JSON Response Schema</a>

The response for a 'single entry' endpoint is the same as for 'entry listing'
endpoint responses, except that the value of the `data` field MUST have only one or zero entries.

> In the default `json` response format, this means the value of the `data` field MUST be

Reviewer: JSON

The database-provider-specific prefixes currently assigned are listed in the `providers.json` file
provided in the main repository. This file serves as a machine-readable list of OPTiMaDe providers.

> The content of the `providers.json` file complies with the default `json` format specification for API responses.

Reviewer: JSON

Author: All don, thanks @CasperWA
